pkg/controller/appdefinition/networkpolicy.go

```diff
@@ -1,25 +1,235 @@
+	corev1 "k8s.io/api/core/v1"
 	networkingv1 "k8s.io/api/networking/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-func NetworkPolicy(req router.Request, resp router.Response) error {
+// NetworkPolicyForApp creates a single Kubernetes NetworkPolicy that restricts incoming network traffic
+// to all pods in an app, so that they cannot be reached by pods from other projects.
+func NetworkPolicyForApp(req router.Request, resp router.Response) error {
+	cfg, err := config.Get(req.Ctx, req.Client)
+	if err != nil {
+		return err
+	} else if *cfg.DisableNetworkPolicies {
+		return nil
+	}
+
 	app := req.Object.(*v1.AppInstance)
+	appNamespace := app.Namespace        // this is where the AppInstance lives
+	podNamespace := app.Status.Namespace // this is where the app is actually running
 
+	// create the NetworkPolicy for the whole app
+	// this allows traffic only from within the project
 	resp.Objects(&networkingv1.NetworkPolicy{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:      app.Name,
-			Namespace: app.Status.Namespace,
+			Namespace: podNamespace,
+		},
+		Spec: networkingv1.NetworkPolicySpec{
+			PodSelector: metav1.LabelSelector{
+				MatchLabels: labels.Managed(app),
+			},
+			Ingress: []networkingv1.NetworkPolicyIngressRule{{
+				From: []networkingv1.NetworkPolicyPeer{{
+					NamespaceSelector: &metav1.LabelSelector{
+						MatchLabels: map[string]string{
+							labels.AcornAppNamespace: appNamespace,
+						},
+					}},
+				},
+			}},
+			PolicyTypes: []networkingv1.PolicyType{networkingv1.PolicyTypeIngress},
+		},
+	})
+	return nil
+}
+
+// NetworkPolicyForIngress creates Kubernetes NetworkPolicies to allow traffic to exposed HTTP ports on
+// Acorn apps from the ingress controller. If the ingress controller namespace is not defined, traffic from
+// all namespaces will be allowed instead.
+func NetworkPolicyForIngress(req router.Request, resp router.Response) error {
+	cfg, err := config.Get(req.Ctx, req.Client)
+	if err != nil {
+		return err
+	} else if *cfg.DisableNetworkPolicies {
+		return nil
+	}
+
+	ingress := req.Object.(*networkingv1.Ingress)
+	appName := ingress.Labels[labels.AcornAppName]
+	projectName := ingress.Labels[labels.AcornAppNamespace]
+
+	// create a mapping of k8s Service names to published port names/numbers
+	svcNameToPorts := make(map[string][]networkingv1.ServiceBackendPort)
+	for _, rule := range ingress.Spec.Rules {
+		for _, path := range rule.HTTP.Paths {
+			svcName := path.Backend.Service.Name
+			port := path.Backend.Service.Port
+			svcNameToPorts[svcName] = append(svcNameToPorts[svcName], port)
+		}
+	}
+
+	for svcName, ports := range svcNameToPorts {
+		// get the Service from k8s
+		svc := corev1.Service{}
+		err = req.Get(&svc, ingress.Namespace, svcName)
+		if err != nil {
+			// is this okay to do? sometimes the Service doesn't exist yet when we need to get it
+			// so errors might occasionally appear in the logs but things eventually settle and work
+			return err
+		}
```

Comment:
Wanted to leave a quick comment on this for reviewers - is this the right approach here? The service will exist eventually, so I need this request to be re-queued, but I'm not sure if there's another way to do it besides just returning an error.

Reply:
Instead of returning an error, you could set `resp.RetryAfter()` and return `nil` here.

Reply:
Thanks! Didn't know that was a thing but that is perfect for this situation.